### ISSUE TITLE
Very slow compile for tree-map over large number of leaves

### ISSUE THREAD

**Issue by njwfish** [google/jax#16084]
### Description

In my problem I want to run gradient descent over a groups of data of different sizes. There is significant heterogeneity in the group sizes; one group might have 100 datapoints, another might have 1000. I need to run a model on each individual datapoint in each group and then sum the output and push that through a loss with a group outcome, which we can call the group_Y. So for a single group we have ${(X_0, ..., X_{N_g}), Y_g}$ and want to compute `z = (model_fn(X_0),...,model_fn(X_{N_g})` and then `loss_fn(z, Y_g)`. We then need to take the mean of the loss over the groups so we can get a scalar output. The very natural way to do this in jax, at least to me, is to use tree maps and tree reduces:

```
def get_mapped_grad_and_value(
        loss_fn: Callable,
        model_fn: Callable,
) -> Tuple[Callable, Callable]:
    """Returns a function that computes the gradient of the loss function with
    respect to the model parameters.

    Args:
        loss_fn (function): loss function
        model_fn (function): model function
        data (tuple): tuple of 

    Returns:
        tuple: loss function, gradient function
    """
    def mapped_loss(model_params, grouped_Xs, grouped_Ys):
        """Compute the loss function mapped over groups.
        
        Args:
            model_params: model parameters
            grouped_Xs: dict of np.ndarrays of group datapoints
            grouped_Ys: dict of float group outcomes
            
        Returns:
            mean loss over all groups"""
        num_groups = len(grouped_Ys)
        group_losses = jax.tree_util.tree_map(
            lambda group_X, group_Y: 
            loss_fn(model_fn(model_params, group_X), group_Y) / num_groups, 
            grouped_Xs, grouped_Ys
        )
        loss = jax.tree_util.tree_reduce(lambda x, y: x + y, group_losses)
        return loss
    
    grad_and_loss = jax.jit(jax.value_and_grad(mapped_loss))
    return grad_and_loss
```

Now I do have a large dataset, about 1000-8000 groups, with roughly 1-20 million datapoints partitioned between the groups. But the compile times I am seeing are very long, about 4-5 hours. I am wondering what, if anything, I can do to speed up compile times here.

It's worth noting I am seeing very similar compile times for any analogous code:

```
def mapped_loss(model_params, grouped_Xs, grouped_Ys):
    group_losses = jax.tree_util.tree_map(
        lambda group_X, group_Y, group_N: 
        loss_fn(model_fn(model_params, group_X), group_Y), 
        grouped_Xs, grouped_Ys
    )
    return np.mean(np.array(list(group_losses.values())))
```

```
def mapped_loss(model_params, grouped_Xs, grouped_Ys):
    loss = 0
    for g in grouped_Xs:
        loss += loss_fn(model_fn(model_params, grouped_Xs[g]), grouped_Ys[g])
    return loss

```

Currently I am compiling the loss/gradient of the loss function, tree-mapping that, and using a non-compiled tree-reduce to actually compute the final gradient update. This works _fine_ if a bit slow, but it makes it prohibitive to use more interesting optimizers from e.g. jaxopt which use jitted loops.  

Maybe it is not possible to speed this up, but if it is I'd be grateful. 

### What jax/jaxlib version are you using?

jax v0.4.10

### Which accelerator(s) are you using?

M1 CPU

### Additional system info

Python 3.9.13

**Comment by jakevdp** [comment]
This is expected, and there's not really any way around it. `tree_map` is essentially a host-side loop, and when you JIT-compile large host-side loops you will generally observe very large compilation times.

The fix would be to avoid tree-map, and instead use direct array operations, though admittedly in situations like those you describe, it can be difficult to express. In your case, where there are a variable number of elements, I'd use a single array padded to the maximum size of the elements, rather than writing a tree map over arrays of different sizes.

Hope that helps!

**Comment by njwfish** [comment]
Yes that makes sense and that was my expectation. 

I guess one thing I might ask is whether it is possible to do a "weak" compilation in settings like this so that these complex mapped functions can be passed into other jitted functions? So there would be no or almost no benefits of compilation but the compilation would be fast. That would be very useful in this project (and would have been useful in some others I've worked on in the past).

**Comment by jakevdp** [comment]
Can you say more about this weak compilation idea? I'm not sure I fully understand.

**Comment by patrick-kidger** [comment]
If you're on the CPU then you can try wrapping your `model_fn` in an `equinox.internal.noinline`. (See [here](https://github.com/patrick-kidger/equinox/tree/main/equinox/internal).) This will compile the wrapped function only once for each group size. The downside is that this will probably only work on the CPU, and it dramatically increases runtime.

The "real" fix here is indeed as Jake says -- pad out the groups to just a few sizes, and then replace multiple calls `fn(group1); fn(group2)` to groups of the same size with a `jax.vmap(fn)(jnp.concatenate([group1, group2])`.

**Comment by njwfish** [comment]
Yes! Thank you @patrick-kidger I think this noinline functionality is basically doing what I was thinking of re: "weak compilation". 

More generally I guess I think this behavior is odd, and I think it'd be quite nice to have an override (e.g. a parameter in jit that would let something act `jitted` for the purpose of use in other functions/recompilation without inlining every instance which I presume is what's exploding the compile times).  The reason I think it's odd is that in my use case I can do the following:

```
def get_mapped_loss_and_grad(
        loss_fn: Callable,
        model_fn: Callable,
) -> Tuple[Callable, Callable]:
    """Returns a function that computes the gradient of the loss function with
    respect to the model parameters.

    Args:
        loss_fn (function): loss function
        model_fn (function): model function
        data (tuple): tuple of 

    Returns:
        tuple: loss function, gradient function
    """

    _loss_and_grad_fn = jax.jit(jax.value_and_grad(loss_fn))

    def loss_and_grad_fn(model_params, group_Xs, group_Ys):
        num_groups = len(group_Ys)
        group_losses = jax.tree_util.tree_map(
            lambda group_X, group_Y, group_N: 
            _loss_and_grad_fn(model_params, group_X, group_Y), 
            group_Xs, group_Ys
        )
        mean_loss, grad = jax.tree_util.tree_reduce(
            lambda x, y: (x[0] + y[0], x[1] + y[1]), group_losses, 
            is_leaf=lambda x: not isinstance(x, dict)
        )
        return mean_loss, grad

    return loss_and_grad_fn

```

The compile time is a couple minutes to compile in total to run this with all my groups. The code is also very fast!

But if I run `loss_and_grad_fn = jax.jit(loss_and_grad_fn)` (or I compile any of the similar functions outlined above) that takes many hours to compile for qualitatively very little actual gain in runtime (I can actually run tests for this if it would be useful/interesting). And not being able to jit comile is annoying! Lots of code presumes you can jit internal functions almost without cost. Now presumably as Jake and Patrick suggest the general solution is to pad things out in some appropriate way. But this gets quite cumbersome to implement and understand (if you aren't familiar with jax already). I think it would also be quite wasteful of memory which is a bottleneck in my problem. 

Now in my particular case I can re-write my codebase to use jax.experimental.sparse (I knew this from the start but was intentionally trying to steer clear of experimental code), so I will probably go that route! I guess I just want to gesture vaguely that I think it would be reasonable to support something like this allowing users to more explicitly tradeoff between compile time and run time in their specific setting with some options on jit like those Patrick is pointing to in equinox. On the other hand, maybe equinox is all I need and this doesn't need to make its way into the base package.

Anyway, this has been quite helpful! Thank you all for the discussion. If folks feel this is resolved I'm happy to close the issue.

**Comment by njwfish** [comment]
Hello, I wanted to circle back here. I have re-implemented this code using jax.experimental.sparse, but it is actually 1) much slower and 2) much less memory efficient than my very hacky solution above where I compile each group independently. I find this pretty weird behavior.

Essentially my problem is like this: I have a set of matrices with different numbers of rows X_1,...,X_n. I push these through a model giving a set of vectors of different lengths. My loss function is defined as a function of the sums of these vectors and an additional scalar Y_i associated with each X_i matrix. Since my loss is a function of a sum in this way I can use a huge sparse matrix to compute the sums over the groups in a jax shape-friendly way. 

Maybe I am setting up the sparse matrix problem in the wrong way? I can also produce a MWE if that would help figure things out here! I find it puzzling that it is both much slower and more memory intense than the group-wise alternative especially given that the group-wise code cannot even be jitted.

**Comment by njwfish** [comment]
Okay I put together a minimal working example which I think demonstrates this issue very clearly, colab notebook [here](https://colab.research.google.com/drive/1x0rLtaNyjmA8YdQSvTAlvSwpEb7qWXJq?usp=sharing).

This shows that in the setting I describe above (which may be clearer to understand in the data generating process cell in the notebook) we have:
1. faster evaluation of the sparse gradient when not vmapping, but slower hessian evaluation.
2. significantly slower and much more memory hungry gradient (4x slower) and hessian (8x slower) evaluation of sparse code than the tree_mapped counterpart when vmapping.

These results are very surprising given that the tree_mapped code cannot be jitted collectively and has to be computed for each group separately and then combined. Also the tree_map solution just feels very hacky. I would expect the sparse implementation to handily beat the tree_map solution. These high-level results remain true if I swap out the BCOO matrix for BCSR or if I compile the G matrix with constant folding.

@jakevdp I would be very interested in advice or thoughts on what is making the sparse version so slow here and whether it is fixable either in my example or on the backend. Thanks so much!

**Comment by jakevdp** [comment]
What is `get_mapped_fn`? It's not defined in the notebook you linked to.

**Comment by njwfish** [comment]
Sorry it is defined there, it is just hidden in the notebook, under "Helper Functions for Tree Approach".

```
#@title Helper Functions for Tree Approach
def reduce_sum(tree):
    """Reduce sum of tree.
    Args:
        tree: tree to sum.
    Returns:
        float: sum of tree.
    """
    return jax.tree_util.tree_reduce(
        jax.jit(lambda x, y: x + y if isinstance(x, np.ndarray) else [xi + yi for (xi, yi) in zip(x, y)]),
        tree, is_leaf=lambda x: not isinstance(x, dict)
    )

def get_mapped_fn(fn):
    """Get mapped function.
    Args:
        fn: function to map.
    Returns:
        function: mapped function."""
    def mapped_fn(model_params, args):
        """Map function over tree, taking model_params as first argument and summing results.
        Args:
            model_params: model parameters.
            args: arguments to pass to fn.
        Returns:
            float: sum of mapped function.
        """
        group_values = jax.tree_util.tree_map(
            lambda *args: fn(model_params, *args), *args
        )
        return reduce_sum(group_values)
    return mapped_fn
```

**Comment by jakevdp** [comment]
Thanks - I couldn't find it via cmd-F, but I see it now that it's un-minimized.

**Comment by jakevdp** [comment]
Sorry, there's a lot there and I'm not sure exactly what I should be looking at. Is your core question still why the tree-mapped approach leads to long compile times?

**Comment by njwfish** [comment]
Apologies, yep theres a couple things going on. My core question is actually: why, especially when vmapped, is the sparse code _slower_ than the tree-mapped version? And maybe more important, why is the sparse code so intensely memory-hungry?

To me, for all the reasons discussed above, it feels like the tree-map solution should be strictly worse but it seems to be much better in this situation. I am wondering if its an implementation issue on my end or something strange in the sparse implementation.

**Comment by jakevdp** [comment]
In general, when comparing implementations based on `vmap` with implementations based on Python loops (which is effectively what `tree_map` is), you should expect `vmap` to have faster compile times but slower runtimes, and you should expect Python loops to have slower compile times and faster runtimes.

Why? Basically `vmap` just sends a single buffer and a single operation to XLA: it's very few instructions, so the compiler has very few decisions to make about the execution strategy. This leads to fast compilation.

On the other hand, with a loop-based implementation, you're sending a long sequence of operations to the compiler. This leads to long compile times, because there are many decisions to make about the best order/fusing of operations. But the flip side of long compilation is that the compiler can produce a very well-optimized kernel, and you'll often find that runtimes are faster.

Those are general principles that don't always hold exactly in any particular situation, but I suspect it's relevant to understanding what's happening in your case.

**Comment by jakevdp** [comment]
Regarding sparse operations with `jax.experimental.sparse`, I suspect it's slow because that experimental package as a whole is very slow. It's effectively a reference implementation of sparse operations on top of XLA, which is not well-suited to sparse operations, and so you should generally expect it to be slow.

**Comment by njwfish** [comment]
Ah if the sparse package is just expected not to be that performant then I think that answers my follow-up question.
